pages/learn/deploy/offline-updates.md

```diff
@@ -0,0 +1,276 @@
+
+$ rm ${tmpimg}
+```
+
+### Process of Reprovisioning
+
+> Warning: For devices with internal storage, this procedure erases the internal media of your device. Hence, remove any mass storage devices containing user data.
+
+With the update media ready having the latest release of the application preloaded. Follow the device's provisioning instructions present on balenaCloud dashboard for your specific device.
+
+For example: For Raspberrypi devices, insert the recently flashed SD card and power up the device. When the process is complete, (re)connect any mass storage devices containing user data back to the device. Reconnect the device to the local air-gapped network(s). Later, use SSH to connect and inspect application logs, etc.
+
+#### Strategies to remotely update with an SD card or USB device
+
+If a device isn't locally deployed, one can ship the flashed SD cards or USB sticks/drives to a remote location. There someone can run the update by simply inserting them into the devices and booting.
+
+If the target device exists on an air-gapped or Internet restricted network, inserting [ssh](https://www.balena.io/docs/reference/OS/configuration/#sshkeys) keys during the configuration step will allow fleet managers at the remote site to connect into the device directly via OpenSSH and verify the update by examining container logs, etc.
+
+If the target device is connected via a low-bandwidth connection, it should eventually establish a connection to balenaCloud. Depending on the connection's quality, it may respond to [Web Terminal][web-terminal] commands and output container logs to the dashboard.
+
+### Update Device Registration(s)
+
+This section deals with device registration in the cloud (balenaCloud or openBalena). The following steps help in manually reflecting the new state of the device for offline devices or devices on private networks without Internet access.
+
+#### Patch State Endpoint - Update Device State in the Cloud
+
+This step updates the device state in the cloud (balenaCloud or openBalena)
+on behalf of the device, as if the device was online. It is only required for
+devices that do not have an internet connection. If the device has an internet
+connection, even if low bandwidth, this step may be skipped because the device
+itself will be able to contact the cloud to update its state.
+
+```bash
+$ os_version_tag=$(echo ${os_version} | awk -F'+' '{print $1}')
+$ os_version_semver=$(echo ${os_version_tag} | sed 's/v//g')
+$ os_revision=$(echo ${os_version} | awk -F'+' '{print $2}' | awk -F'.' '{print $1}')
+$ os_variant=$(echo ${os_version} | awk -F'+' '{print $2}' | awk -F'.' '{print $2}')
+
+$ supervisor_version=$(curl --silent https://raw.githubusercontent.com/balena-os/meta-balena/${os_version_tag}/meta-balena-common/recipes-containers/resin-supervisor/resin-supervisor.inc \
+    | grep SUPERVISOR_TAG \
+    | awk '{print $3}' \
+    | sed 's/"//g' \
+    | sed 's/v//g')
+
+# (e.g.) staging
```

Comment:
Instead of this `# (e.g.) staging` comment, how about replacing the two occurrences of `balena-staging.com` with an environment variable? The CLI uses the `BALENARC_BALENA_URL` environment variable to select a different cloud domain name (openBalena, staging, ...) than the `balena-cloud.com` default, so I suggest:
```
if [ -z "$BALENARC_BALENA_URL" ]; then
    BALENARC_BALENA_URL=balena-cloud.com
fi

$ release_id=$(curl --silent \
    "https://api.${BALENARC_BALENA_URL}/v6/release?
...

$ curl --silent \
    -X 'PATCH' "https://api.${BALENARC_BALENA_URL}/v6/device(uuid='${uuid}')" \
...
```
(Yes, ideally this bash code would be replaced with a new flag for the `balena preload` command.)

Reply:
The only problem with this @pdcastro is that it doesn't fit into the overall command structure we have going on in the document. Everything is a shell command and here comes a bash script IF statement. I can make this into a one-liner though.

`if [ -z "$BALENARC_BALENA_URL" ]; then  BALENARC_BALENA_URL=balena-cloud.com; fi`

Reply:
Replacing stuff with this.

Reply:
@pdcastro Do you think its a good idea to add somewhere that this all is experimental as the team will greatly smoothen out the entire process as things come along. I think its worth it to have it somewhere as the blog does it too. 

cc: @ab77